zerver/webhooks/bitbucket3/view.py

```diff
@@ -0,0 +1,161 @@
+    event_type = change["type"]
+    repo_name = payload["repository"]["name"]
+    tag_name = change["ref"]["displayId"]
+    user_name = payload["actor"]["name"]
+
+    if event_type == "ADD":
+        action = "pushed"
+    elif event_type == "DELETE":
+        action = "removed"
+    else:
+        raise UnexpectedWebhookEventType("BitBucket3", event_type)
+
+    subject = BITBUCKET_TOPIC_TEMPLATE.format(repository_name=repo_name)
+    body = get_push_tag_event_message(
+        user_name,
+        tag_name,
+        action=action)
+    return {"subject": subject, "body": body}
+
+def repo_push_handler(payload: Dict[str, Any], branches: Optional[str]=None) -> List[Dict[str, str]]:
+    data = []
+    for change in payload["changes"]:
+        event_target_type = change["ref"]["type"]
+        if event_target_type == "BRANCH":
+            branch = change["ref"]["displayId"]
+            if branches:
+                if branch not in branches:
+                    continue
+            data.append(repo_push_branch_data(payload, change))
+        elif event_target_type == "TAG":
+            data.append(repo_push_tag_data(payload, change))
+        else:
+            raise UnexpectedWebhookEventType("BitBucket3", event_target_type)
+    return data
+
+EVENT_HANDLER_MAP = {
+    "repo:comment:added": partial(repo_comment_handler, action="commented"),
+    "repo:comment:edited": partial(repo_comment_handler, action="edited their comment"),
+    "repo:comment:deleted": partial(repo_comment_handler, action="deleted their comment"),
+    "repo:forked": repo_forked_handler,
+    "repo:modified": repo_modified_handler,
+    "repo:refs_changed": repo_push_handler,
+    "pr:comment:added": None,
+    "pr:comment:edited": None,
+    "pr:comment:deleted": None,
+    "pr:declined": None,
+    "pr:deleted": None,
+    "pr:merged": None,
+    "pr:modified": None,
+    "pr:opened": None,
+    "pr:reviewer:approved": None,
+    "pr:reviewer:needs_work": None,
+    "pr:reviewer:updated": None,
+    "pr:reviewer:unapproved": None,
+}  # type Dict[str, Optional[Callable[..., List[Dict[str, str]]]]]
+
+def get_event_handler(eventkey: str) -> Callable[..., List[Dict[str, str]]]:
+    # The main reason for this function existance is because of mypy
+    handler = EVENT_HANDLER_MAP.get(eventkey)  # type: Any
+    if handler is None:
+        raise UnexpectedWebhookEventType("BitBucket3", eventkey)
+    return handler
+
+@api_key_only_webhook_view("Bitbucket3")
+@has_request_variables
+def api_bitbucket3_webhook(request: HttpRequest, user_profile: UserProfile,
+                           payload: Dict[str, Any]=REQ(argument_type="body"),
+                           branches: Optional[str]=REQ(default=None),
+                           user_specified_topic: Optional[str]=REQ("topic",
+                                                                   default=None)
+                           ) -> HttpResponse:
```

Comment:
Hmm.. this isn't a huge deal but the arguments in the function signature look a bit awkward here. Could you please change them to be:

```python
@has_request_variables
def api_bitbucket3_webhook(
        request: HttpRequest, user_profile: UserProfile,
        payload: Dict[str, Any]=REQ(argument_type="body"),
        branches: Optional[str]=REQ(default=None),
        user_specified_topic: Optional[str]=REQ("topic", default=None)
) -> HttpResponse:
    eventkey = payload.get('eventKey')
```

Reply:
Agreed, it did look pretty ugly. I've made the change! :)